### PR TITLE
Prepare for chaining autogenerated conversion methods 

### PR DESCRIPTION
**Pull request by wojtek-t**
Part of #6800

This PR is prerequisite for chaining autogenerated conversion methods. The next steps are:
- test checking whether auto-generated code wasn't manually edited
- call auto-generated methods one from another [this is easy modulo effective handling of defaulting functions]

@lavalamp @smarterclayton @fgrzadkowski

### CODE REVIEW

**Comment by wojtek-t** [comment]
@lavalamp @smarterclayton 
To be clear on calling auto-generated methods one from another. The issue with it is that defaulting functions, are performed on reflections:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/conversion/converter.go#L462
So far, I have no idea how to avoid reflections for that, so if you have any suggestions please let me know.

**Comment by smarterclayton** [comment]
In this change, please add instructions on generation to `docs/api_changes.md` so that folks needing to add things to the API (@liggitt) have some place to go.

**Comment by wojtek-t** [comment]
>
In this change, please add instructions on generation to docs/api_changes.md so that folks needing to add things to the API (@liggitt) have some place to go.

I've started writing this doc and uploaded few lines of it. However, I might not be able to finish it today. So it would be great if you could take a look at the code, before I'm done with the document.

**Comment by smarterclayton** [comment]
Separate the conversion changes into their own commit so it's easier to review.  We probably want to formalize that in the policy around handling these (always put generations in a separate commit).

**Comment by wojtek-t** [comment]
>
Separate the conversion changes into their own commit so it's easier to review. We probably want to formalize that in the policy around handling these (always put generations in a separate commit).

Done

**Comment by smarterclayton** [comment]
Instead of 

```
	err := newer.Scheme.AddGeneratedConversionFuncs(
		func(in *AWSElasticBlockStoreVolumeSource, out *newer.AWSElasticBlockStoreVolumeSource, s conversion.Scope) error {
			return convertV1beta3AWSElasticBlockStoreVolumeSourceToApiAWSElasticBlockStoreVolumeSource(in, out, s)
		},
```

do 

```
	err := newer.Scheme.AddGeneratedConversionFuncs(
		convertV1beta3AWSElasticBlockStoreVolumeSourceToApiAWSElasticBlockStoreVolumeSource,
```

**Comment by wojtek-t** [comment]
I'm going to improve how the cmd/kube-conversion/conversion.go is used in this PR too (tomorrow), but this won't change the code that is generated. So it would be great if you could look into that. Thanks!

**Comment by smarterclayton** [comment]
Instead of `convertV1beta3AWSElasticBlockStoreVolumeSourceToApiAWSElasticBlockStoreVolumeSource` can you do `convert_v1beta3_AWSElasticBlockStoreVolumeSource_To_internal_AWSElasticBlockStoreVolumeSource`

**Comment by smarterclayton** [comment]
What's the golang limit on function names?

File: docs/api-changes.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+API Changes
```
Comment (thockin) [inline docs/api-changes.md]: we already have a docs/devel/api_changes.md - merge?

Comment (wojtek-t) [inline docs/api-changes.md]: Thanks @thockin - sounds good to me. I will merge those two files (most probably tomorrow).

File: pkg/conversion/generator.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ import (
 type Generator interface {
 	GenerateConversionsForType(version string, reflection reflect.Type) error
 	WriteConversionFunctions(w io.Writer) error
+	WriteConversionFunctionsForRegister(w io.Writer) error
```
Comment (lavalamp) [inline pkg/conversion/generator.go]: What does "ForRegister" mean?

Comment (wojtek-t) [inline pkg/conversion/generator.go]: This wasn't the best name. Fixed.

**Comment by wojtek-t** [comment]
>
Instead of `convertV1beta3AWSElasticBlockStoreVolumeSourceToApiAWSElasticBlockStoreVolumeSource` can you do `convert_v1beta3_AWSElasticBlockStoreVolumeSource_To_internal_AWSElasticBlockStoreVolumeSource`

I think that names without underscores ale closer to the styleguide (I didn't see any names with underscores anywhere). Why do you want to change this?

All the other comments are applied - PTAL

**Comment by smarterclayton** [comment]
> On Apr 29, 2015, at 4:57 AM, Wojciech Tyczynski <notifications@github.com> wrote:
> 
> Instead of convertV1beta3AWSElasticBlockStoreVolumeSourceToApiAWSElasticBlockStoreVolumeSource can you do convert_v1beta3_AWSElasticBlockStoreVolumeSource_To_internal_AWSElasticBlockStoreVolumeSource
> 
> I think that names without underscores ale closer to the styleguide (I didn't see any names with underscores anywhere). Why do you want to change this?
> 
The style guide really doesn't apply to generated functions.  The fully transformed version is almost unreadable, so the underscores bring it back.

I'd be willing to make an exception in our style guide to improve the readability and scanability of the list.  Also, it preserves the original casing of the names which is probably better.

**Comment by wojtek-t** [comment]
>
The style guide really doesn't apply to generated functions.  The fully transformed version is almost unreadable, so the underscores bring it back.
I'd be willing to make an exception in our style guide to improve the readability and scanability of the list.  Also, it preserves the original casing of the names which is probably better.

I don't fully agree with that, but I see your point. Changed.

PTAL

**Comment by wojtek-t** [comment]
BTW - I have almost ready the next PR, that is calling those conversion functions one from another without using reflections. That PR gives us another ~75% improvements (which is more than 10x from the original results), but since that new PR depends on this one, I will wait with submitting that one before this one is merged.

**Comment by smarterclayton** [comment]
One more nag - please alpha sort the conversion function references in 

```
newer.Scheme.AddGeneratedConversionFuncs(
convert_v1beta3_AWSElasticBlockStoreVolumeSource_To_api_AWSElasticBlockStoreVolumeSource,
```

So you can quickly binary search visually to find them.  Will make it easier to identify missed conversion bugs and diffs.  Best order is a forced order.

**Comment by wojtek-t** [comment]
>
One more nag - please alpha sort the conversion function references in 
```
newer.Scheme.AddGeneratedConversionFuncs(
convert_v1beta3_AWSElasticBlockStoreVolumeSource_To_api_AWSElasticBlockStoreVolumeSource,
```
So you can quickly binary search visually to find them.  Will make it easier to identify missed conversion bugs and diffs.  Best order is a forced order.

Done - PTAL

**Comment by smarterclayton** [comment]
This is LGTM barring any final review from @lavalamp

**Comment by wojtek-t** [comment]
Thanks @smarterclayton !

**Comment by lavalamp** [comment]
My feedback is that it's not obvious which sections of the file are autogenerated and which aren't (add comments before the functions?) and that maybe you could generate it for v1, too, to keep that in sync?

**Comment by lavalamp** [comment]
But I don't think that needs to block the PR.

**Comment by wojtek-t** [comment]
>
My feedback is that it's not obvious which sections of the file are autogenerated and which aren't (add comments before the functions?) 

I definitely agree with it - tomorrow I will be working on the test whether the generated code wasn't manually edited and I will add it as part of it.

>
and that maybe you could generate it for v1, too, to keep that in sync?

Can I do it in a separate PR?

>
But I don't think that needs to block the PR.

If you're fine with merging as is, that would be great. I will address your two comments in a separate PR tomorrow.

**Pull request merged**